### PR TITLE
Add TypoScript support

### PR DESCRIPTION
**Pull request by andreaswolf**

### CODE REVIEW

**Comment by andreaswolf** [comment]
One thing to solve here: how should a single Fractor communicate its result back to the caller? There is no clear rule for that right now, so @simonschaufi and me discussed that today. We came to the following conclusion:

* null = don't change anything
* a node or a list of nodes = replace the current node with that
* an integer = do something else (specific to each file type?), e.g. remove the node, stop traversal of the node (and possibly its children)

This is at least similar to what Rector does, probably it's 100% the same; I did not check that again for now.

Compared to the XML Fractor, an additional challenge with TS is that we need to return a list of nodes from the iterator, since there is no document node that will automatically collect all changed nodes (either directly or implicitly in its children). This needs to be solved, but is not a big deal.

Deciding if a node can be handled and recording the changes and who changed what is part of the abstract base Fractor, so it must basically be reimplemented for each file type. That's not easy to change, so we accept that for now (and it leaves us more leeway to adapt things for file types where it is needed).

File: packages/fractor-typoscript/composer.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
+{
+    "name": "a9f/fractor-typoscript",
+    "description": "TypoScript extension for the File Read-Analyse-Change TOol. Allows modifying TYPO3 TypoScript configuration files",
```
Comment (simonschaufi) [inline packages/fractor-typoscript/composer.json:3]: Tool

File: packages/fractor-typoscript/config/application.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+return static function (ContainerConfigurator $containerConfigurator, ContainerBuilder $containerBuilder): void {
+    $services = $containerConfigurator->services();
+    $services->defaults()
+        ->autowire()
+        ->autoconfigure();
+
+    $services->load('a9f\\FractorTypoScript\\', __DIR__ . '/../src/');
+
+    $services->set(Helmich\TypoScriptParser\Tokenizer\Tokenizer::class);
```
Comment (simonschaufi) [inline packages/fractor-typoscript/config/application.php:17]: import classes

File: packages/fractor-typoscript/src/TypoScriptStatementsIterator.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        $nodeRemoved = false;
+        foreach ($this->visitors as $visitor) {
+            $result = $visitor->enterNode($node);
+        }
+
+        foreach ($this->visitors as $visitor) {
+            $visitor->leaveNode($node);
+        }
+    }
+}
```
Comment (simonschaufi) [inline packages/fractor-typoscript/src/TypoScriptStatementsIterator.php:60]: eof missing

File: packages/fractor-typoscript/tests/Fixture/StatementCollectingVisitor.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    /**
+     * @param list<Statement> $statements
+     */
+    public function afterTraversal(array $statements): void
+    {
+        $this->calls[] = sprintf('%s:afterTraversal:%s', $this->visitorName, count($statements));
+    }
+};
```
Comment (simonschaufi) [inline packages/fractor-typoscript/tests/Fixture/StatementCollectingVisitor.php:46]: eof missing

File: packages/fractor-typoscript/tests/Fixture/StatementCollectingVisitor.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+declare(strict_types=1);
+
+namespace a9f\FractorTypoScript\Tests\Fixture;
+
+use a9f\FractorTypoScript\Contract\TypoScriptNodeVisitor;
+use Helmich\TypoScriptParser\Parser\AST\Statement;
+
+final class StatementCollectingVisitor implements TypoScriptNodeVisitor {
```
Comment (simonschaufi) [inline packages/fractor-typoscript/tests/Fixture/StatementCollectingVisitor.php:10]: opening brace on new line

**Review by simonschaufi — CHANGES_REQUESTED** [review]

**Comment by andreaswolf** [comment]
@simonschaufi I just resolved all remarks that will be fixed by our pipeline. No need to comment those :)

File: packages/fractor-typoscript/src/TypoScriptFileProcessor.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    public function __construct(
+        private iterable $rules,
+        private Parser $parser,
+        private PrettyPrinter $printer
+    ) {
+    }
+
+    public function canHandle(File $file): bool
+    {
+        return in_array($file->getFileExtension(), ['typoscript', 'ts']);
```
Comment (simonschaufi) [inline packages/fractor-typoscript/src/TypoScriptFileProcessor.php:32]: What about user and page ts config? I think they have their own file extension sometimes

File: packages/typo3-fractor/rules/TYPO3v10/TypoScript/RemoveUseCacheHashFromTypolink.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+use a9f\FractorTypoScript\TypoScriptStatementsIterator;
+use Helmich\TypoScriptParser\Parser\AST\Operator\Assignment;
+use Helmich\TypoScriptParser\Parser\AST\Statement;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/10.0/Deprecation-88406-SetCacheHashnoCacheHashOptionsInViewHelpersAndUriBuilder.html
+ */
+final class RemoveUseCacheHashFromTypolink extends AbstractTypoScriptFractor
```
Comment (simonschaufi) [inline packages/typo3-fractor/rules/TYPO3v10/TypoScript/RemoveUseCacheHashFromTypolink.php:17]: Fractor suffix missing

File: packages/fractor-typoscript/src/TypoScriptFileProcessor.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+        $output = new BufferedOutput();
+        $this->printer->setPrettyPrinterConfiguration(PrettyPrinterConfiguration::create() ->withEmptyLineBreaks());
+        $this->printer->printStatements($statements, $output);
+        $file->changeFileContent($output->fetch());
+    }
+
+    public function allowedFileExtensions(): array
+    {
+        return ['typoscript', 'ts'];
```
Comment (simonschaufi) [inline packages/fractor-typoscript/src/TypoScriptFileProcessor.php:50]: same here

**Comment by simonschaufi** [comment]
A highly requested feature is missing: formatting the result with space or tab indentation. Can this be configured or is this not possible?

File: packages/fractor-typoscript/src/TypoScriptFileProcessor.php (hunk 6 above)
Comment (andreaswolf) [reply packages/fractor-typoscript/src/TypoScriptFileProcessor.php:32]: Yeah, and there is ext_typoscript_setup.txt as well. I'll add tsconfig for now and a comment that this should be configurable in the future.

**Comment by andreaswolf** [comment]
> A highly requested feature is missing: formatting the result with space or tab indentation. Can this be configured or is this not possible?

Should be possible, I'll have a look.

**Comment by andreaswolf** [comment]
> A highly requested feature is missing: formatting the result with space or tab indentation. Can this be configured or is this not possible?

I would open an issue and postpone that for now, so we can get this merged now.

**Review by simonschaufi — APPROVED** [review]

**Pull request merged**